src/main/scala/epic/logo/Inferencer.scala

```diff
@@ -0,0 +1,7 @@
+package epic.logo
+
+trait Inferencer[S] {
+  def initialState: S
+  def reduceStates(state1: S, state2: S): S
```

Comment:
Arbitrary state that the implementers of inferencers may want to carry around. The abstraction is like mapreduce: you return some stuff, and also say how to combine that stuff, then you get the combined stuff in various callbacks. Allows inferencers to have internal state without worrying about threadsafety. Sounds like I need some comments.